perf_bench_test.go

```diff
@@ -24,127 +24,619 @@ package main
+	funcNames    = flag.String("funcNames", "helloworld", "Names of the functions to benchmark, separated by comma")
+	isBindSocket = flag.Bool("bindSocket", false, "Bind all VMs to socket 1 and profile one physical core only "+
+		"(only compatible with a 2x 16-core machine or above)")
+
+	// arguments work for TestProfileSingleConfiguration only
+	vmNum     = flag.Int("vm", 2, "TestProfileSingleConfiguration: The number of VMs")
+	targetRPS = flag.Int("rps", 10, "TestProfileSingleConfiguration: The target requests per second")
+
+	// arguments work for TestProfileIncrementConfiguration only
+	vmIncrStep = flag.Int("vmIncrStep", 4, "TestProfileIncrementConfiguration: The increment VM number")
+	maxVMNum   = flag.Int("maxVMNum", 100, "TestProfileIncrementConfiguration: The maximum VM number")
+
+	// profiler arguments
+	profilerLevel    = flag.Int("l", 1, "Profile level")
+	profilerInterval = flag.Uint64("I", 500, "Print count deltas every N milliseconds")
+	profilerMetrics  = flag.String("metrics", "", "Include or exclude nodes (with "+
+		"+ to add, "+
+		"-|^ to remove, "+
+		"comma separated list, wildcards allowed, "+
+		"add * to include all children/siblings, "+
+		"add /level to specify highest level node to match, "+
+		"add ^ to match related siblings and metrics, "+
+		"start with ! to only include specified nodes)")
 )
 
-func TestBenchRequestPerSecond(t *testing.T) {
+// TestProfileIncrementConfiguration loads requests to VMs and increments VM number after loads start to violate time latency constraint.
+// It also profile counters and RPS at each step. After iteration finishes, it saves results in bench.csv under benchDir folder and
+// plots each counters which are also saved under benchDir folder
+func TestProfileIncrementConfiguration(t *testing.T) {
+	var (
+		idx, rps      int
+		pinnedFuncNum int
+		startVMID     int
+		servedTh      uint64
+		isSyncOffload bool = true
+		metrFile           = "bench.csv"
+		images             = getImages(t)
+		cores              = runtime.NumCPU()
+		metrics            = make([]map[string]float64, *maxVMNum / *vmIncrStep)
+	)
+	log.SetLevel(log.InfoLevel)
+
+	validateRuntimeArguments(t)
+
+	createResultsDir()
+
+	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
+
+	for vmNum := *vmIncrStep; vmNum <= *maxVMNum; vmNum += *vmIncrStep {
+		if vmNum < cores {
+			rps = calculateRPS(vmNum)
+		} else {
+			rps = calculateRPS(cores)
+		}
+
+		log.Infof("vmNum: %d, Target RPS: %d", vmNum, rps)
+
+		bootVMs(t, images, startVMID, vmNum)
+		metrics[idx] = loadAndProfile(t, images, vmNum, rps, isSyncOffload)
+		startVMID = vmNum
+		idx++
+	}
+
+	dumpMetrics(t, metrics, metrFile)
+	profile.PlotCVS(*vmIncrStep, *benchDir, metrFile, "the number of VM")
+
+	tearDownVMs(t, images, startVMID, isSyncOffload)
+}
+
+// TestProfileSingleConfiguration loads requests to fixed number of VMs until loads start to violate tail latency constraint
```

Comment:
these lines are too long

Reply:
loads -> issues
to a fixed number of VMs
until requests start to 
constraint -> threshold
benchDir is a runtime argument

Reply:
Done